### DISCUSSION TITLE
Mastodon v4.1.9 -> v4.2.0 upgrade stuck compilations errors

### DISCUSSION THREAD

**Discussion by coreice** [mastodon/mastodon#27087] [Q&A] [answered]
Hi all,

I am trying to upgrade my Mastodon v4.1.9 instance to v4.2.0 but whatever I am trying, I got stuck on the following error:

```
mastodon@ubuntu:~/live$ RAILS_ENV=production bundle exec rails assets:precompile
Compiling...
Compilation failed:
`isModuleDeclaration` has been deprecated, please migrate to `isImportOrExportDeclaration`
    at isModuleDeclaration (/home/mastodon/live/node_modules/babel-plugin-lodash/node_modules/@babel/types/lib/validators/generated/index.js:2740:35)
    at PluginPass.Program (/home/mastodon/live/node_modules/babel-plugin-lodash/lib/index.js:102:44)
```

My v4.1.9 installation was fairly regular. I started from v4.1.6 which I installed via the documentation on joinmastodon.org some time ago. I upgraded to v4.1.9 in preparation. For the upgrade to v4.2.0 I followed the instructions in the [release notes ](https://github.com/mastodon/mastodon/releases/tag/v4.2.0). After the error occurred (and I ended up with a non-working Mastodon instance) I tried the following:

* Did an `rm -rf` on the `node_modules` and rerun the `bundle install`, `yarn install --frozen-lockfile` and precompilation command
* A `yarn install` without the `--frozen-lockfile` flags
* Run the cleanup commands like `RAILS_ENV=production bundle exec rake tmp:cache:clear` but also `bundle exec rails assets:clean` and `bundle exec rails assets:clobber` I found in other support tickets in this repository.

I am a bit lost now. To provide some information about my system and software:

* yarn: v1.22.19
* node: v16.20.2
* rbenv: v3.2.2
* git status: HEAD detached at: v4.2.0
* ubuntu: 22.04.3 LTS

Any tips are welcome! If you want more information please let me know.

**Comment by coreice** [comment, marked as answer]
Did some further research and I noticed sometimes during the compilation I got kicked out of the SSH session with the server. This kind of triggered me and I went searching in the `syslog` and found entries like this:

```
Sep 23 12:44:30 ubuntu kernel: [62295.396590] Out of memory: Killed process 63987 (node) total-vm:1669416kB, anon-rss:886668kB, file-rss:1156kB, shmem-rss:0kB, UID:1001 pgtables:21204kB oom_score_adj:0
```

Apparently the NodeJS process for building the JavaScript runs out of memory on the system. I am not an expert on this topic but strange compilation errors and behaviour sounds reasonable then. I found discussion https://github.com/mastodon/mastodon/discussions/23171 from earlier this year where user [ClearlyClaire](https://github.com/ClearlyClaire) is suggesting to set the `parallelism` parameter to 1 in the Webpack config. This did the trick for me!

I would say it is a good thing to add this tip to the [Troubleshooting errors](https://docs.joinmastodon.org/admin/troubleshooting/) page on the documentation site. This problems can confuse a lot of people as it is not really clear from the logs what is really happening.

Oh btw: the VM where my Mastodon instance is running on is a 2 GB RAM single core machine so it is (indeed) not that beefy (only a few users are using it so normally it is not a big deal).

**Reply by yaxu** [reply]
Aha thanks, I had the exact same problem.
